### ISSUE TITLE
1.1.329: Regression in composite expressions involving overloads

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#6071]
```python
from typing import Any, Generic, TypeVar, overload

T = TypeVar("T", covariant=True)
T2 = TypeVar("T2", bound=str)


class Foo(Generic[T]):
    ...


@overload
def as_foo(obj: T2) -> Foo[T2]:
    ...

@overload
def as_foo(obj: Any) -> Any:
    ...

def as_foo(obj: Any) -> Any:
    ...


@overload
def from_foo(obj: Foo[T2]) -> T2:
    ...

@overload
def from_foo(obj: Any) -> Any:
    ...

def from_foo(obj: Any) -> Any:
    ...


# Decomposing `as_foo` and `from_foo` calls into two separate lines
# produces the expected type.
foo = as_foo('hi')
reveal_type(from_foo(foo), expected_text="str")


# As of pyright 1.1.329, `from_foo(as_foo(...))` produces the wrong type.
#  Type text mismatch; expected "str" but received "Any"
reveal_type(from_foo(as_foo('hi')), expected_text="str")
```

**Comment by erictraut** [comment]
Thanks for the bug report, and apologies for the regression. Although it may not be obvious, this is the same issue as #6046, which has already been addressed. It will be fixed in the next release of pyright.

By any chance, is your code part of a library whose source is publicly available? If so, we could consider adding it to the  [mypy_primer](https://github.com/hauntsaninja/mypy_primer) list of projects. I run mypy_primer as part of the CI scripts for pyright, and this helps identify any potential regressions before they are merged.

**Comment by rsokl** [comment]
It looks like my [project is already in mypy_primer](https://github.com/hauntsaninja/mypy_primer/blob/896ac85d7a4520ed72cb2f6ff0cd258ee58a7b3a/mypy_primer/projects.py#L828-L834). But perhaps it is not scanning `tests/annotations/`, which is where you would have detected the above regression. I can look into this and open a pr in mypy-primer

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.330, which I just published. It will also be included in an upcoming release of pylance.
